File: question/leetcode/leetcode0049.Group-Anagrams/leetcode0049.go
package leetcode

import "sort"

func groupAnagrams(strs []string) [][]string {
	//key是字符串，值是字符串切片
	//key用于存储排序后的字符串,值则存储原始字符串列表
	mp := map[string][]string{}
	for _, str := range strs {
		//字符串是不可变的,不能直接排序,转成byte slice排序
		s := []byte(str)
		sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
		sortedStr := string(s)
		mp[sortedStr] = append(mp[sortedStr], str)
	}
	ans := make([][]string, 0, len(mp))
	for _, v := range mp {
		ans = append(ans, v)
	}
	return ans
}

File: question/leetcode/leetcode0049.Group-Anagrams/leetcode0049.java
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> mp = new HashMap<>();
        for (String s : strs) {
            char[] cs = s.toCharArray();
            Arrays.sort(cs);
            String sortedStr = new String(cs);
            if (mp.containsKey(sortedStr)) {
                mp.get(sortedStr).add(s);
            } else {
                List<String> list = new ArrayList<>();
                list.add(s);
                mp.put(sortedStr, list);
            }

        }
        List<List<String>> ans = new ArrayList<>();

        for (String k : mp.keySet()) {
            ans.add(mp.get(k));
        }

        return ans;
    }
}